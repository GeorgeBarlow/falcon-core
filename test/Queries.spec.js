// @flow
import Queries from '../src/config/QueryManager';

async function queryFactory(queries, queryCount: number = 1) {
  const array = new Array(queryCount);
  for (let i = 0; i < array.length; i++) {
    array[i] = await queries.add({
      id: `test-id-${i + 1}`,
      type: 'sqlite',
      name: `test-query-${i + 1}`,
      database: '/Users/amila/Desktop/demo.sqlite'
    });
  }
  return Promise.all(array);
}

describe('Queries', function testQueries() {
  beforeEach(async () => {
    this.queries = new Queries();
    await queryFactory(this.queries, 3);
  });

  afterEach(async () => {
    await this.queries.removeAll();
  });

  it('should get all queries', async () => {
    const queries = await this.queries.getAll();
    expect(queries).toMatchSnapshot();
  });

  it('should delete a single query', async () => {
    const queries = await this.queries.getAll();
    const queryIdToDelete = queries[0].id;
    await this.queries.remove(queryIdToDelete);
    const newQueries = await this.queries.getAll();
    expect(newQueries).toMatchSnapshot();
  });

  it('should update a single query', async () => {
    const queries = await this.queries.getAll();
    const queryIdToDelete = queries[0].id;
    await this.queries.update(queryIdToDelete, {
      id: 'test-id-foo',
      name: 'test-query-foo',
      password: 'test-password',
      type: 'sqlite'
    });
    const newQueries = await this.queries.getAll();
    expect(newQueries).toMatchSnapshot();
  });

  it('should get a single query', async () => {
    const queries = await this.queries.getAll();
    const queryId = queries[2].id;
    const query = await this.queries.get(queryId);
    expect(query).toMatchSnapshot();
  });

  it('should perform basic validation', async () => {
    expect(await this.queries.validateBeforeCreation({
      id: 12,
      database: 'aJ@#LJ#@KL$KL@#sdf',
      type: 'sqlite'
    }))
      .toMatchSnapshot();
    expect(await this.queries.validateBeforeCreation({
      id: 12,
      database: '/usr/foo',
      type: 'sqlite'
    }))
      .toMatchSnapshot();
    expect(await this.queries.validateBeforeCreation({
      id: 'foo',
      database: '/usr/foo',
      type: 'sqlite'
    }))
      .toMatchSnapshot();
    expect(await this.queries.validateBeforeCreation({
      id: 'foo',
      name: 'foo',
      database: '/usr/foo',
      type: 'sqlite'
    }))
      .toMatchSnapshot();
    expect(await this.queries.validateBeforeCreation({
      id: 'foo',
      name: 'foo',
      database: '/usr/local/bin/npm',
      type: 'sqlite'
    }))
      .toMatchSnapshot();
  });

  it('should check if a sqlite file is valid or not', async () => {
    expect(await this.queries.validateBeforeCreation({
      id: 'foo',
      name: 'foo',
      database: '/Users/amila/Desktop/demo.sqlite',
      type: 'sqlite'
    }))
      .toEqual({ errorMessages: [], passed: true });
  });
});
