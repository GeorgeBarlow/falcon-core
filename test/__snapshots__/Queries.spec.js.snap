// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries should delete a single query 1`] = `
Array [
  Object {
    "color": "default",
    "id": "test-id-2",
    "name": "test-query-2",
    "query": "SELECT * FROM users",
    "type": "sqlite",
  },
  Object {
    "color": "default",
    "id": "test-id-3",
    "name": "test-query-3",
    "query": "SELECT * FROM users",
    "type": "sqlite",
  },
]
`;

exports[`Queries should get a single query 1`] = `
Object {
  "color": "default",
  "id": "test-id-3",
  "name": "test-query-3",
  "query": "SELECT * FROM users",
  "type": "sqlite",
}
`;

exports[`Queries should get all queries 1`] = `
Array [
  Object {
    "color": "default",
    "id": "test-id-1",
    "name": "test-query-1",
    "query": "SELECT * FROM users",
    "type": "sqlite",
  },
  Object {
    "color": "default",
    "id": "test-id-2",
    "name": "test-query-2",
    "query": "SELECT * FROM users",
    "type": "sqlite",
  },
  Object {
    "color": "default",
    "id": "test-id-3",
    "name": "test-query-3",
    "query": "SELECT * FROM users",
    "type": "sqlite",
  },
]
`;

exports[`Queries should perform basic validation 1`] = `
Object {
  "errorMessages": Array [
    Object {
      "fieldName": "id",
      "message": "\\"id\\" must be a string",
    },
    Object {
      "fieldName": "name",
      "message": "\\"name\\" is required",
    },
    Object {
      "fieldName": "query",
      "message": "\\"query\\" is required",
    },
  ],
  "passed": false,
}
`;

exports[`Queries should perform basic validation 2`] = `
Object {
  "errorMessages": Array [
    Object {
      "fieldName": "id",
      "message": "\\"id\\" must be a string",
    },
    Object {
      "fieldName": "name",
      "message": "\\"name\\" is required",
    },
    Object {
      "fieldName": "query",
      "message": "\\"query\\" is required",
    },
  ],
  "passed": false,
}
`;

exports[`Queries should perform basic validation 3`] = `
Object {
  "errorMessages": Array [
    Object {
      "fieldName": "name",
      "message": "\\"name\\" is required",
    },
    Object {
      "fieldName": "query",
      "message": "\\"query\\" is required",
    },
  ],
  "passed": false,
}
`;

exports[`Queries should perform basic validation 4`] = `
Object {
  "errorMessages": Array [
    Object {
      "fieldName": "query",
      "message": "\\"query\\" is required",
    },
  ],
  "passed": false,
}
`;

exports[`Queries should perform basic validation 5`] = `
Object {
  "errorMessages": Array [
    Object {
      "fieldName": "query",
      "message": "\\"query\\" is required",
    },
  ],
  "passed": false,
}
`;

exports[`Queries should update a single query 1`] = `
Array [
  Object {
    "color": "default",
    "id": "test-id-1",
    "name": "test-query-1",
    "query": "SELECT * FROM users",
    "type": "sqlite",
  },
  Object {
    "color": "default",
    "id": "test-id-2",
    "name": "test-query-2",
    "query": "SELECT * FROM users",
    "type": "sqlite",
  },
  Object {
    "color": "default",
    "id": "test-id-3",
    "name": "test-query-3",
    "query": "SELECT * FROM users",
    "type": "sqlite",
  },
]
`;
